### ISSUE TITLE
🐛 [firestore-bigquery-export] Cross Project Import Script

### ISSUE THREAD

**Issue by javisank** [firebase/extensions#1688]
### [READ] Step 1: Are you in the right place?

Issues filed here should be about a feature request for a **specific extension in this repository**. To file a feature request that affects multiple extensions or the Firebase Extensions platform, please reach out to
[Firebase support](https://firebase.google.com/support/troubleshooter/report/features/) directly.

### [REQUIRED] Step 2: Extension name

This feature request is for extension: firestore-bigquery-export --> fs-bq-import-collection script

### What feature would you like to see?

The extension admits cross project data export. I can have my firestore database in one GCP project and my bigquery instance in another. The "fs-bq-import-collection" script assumes the same projectId for firestore and bigquery. I would like to differenciate those project IDs, so that the bigquery client takes another project ID, and make a complete use of the cross-project functionality of the extension.

### How would you use it?

Another flag in the script, with another prompt in the CLI when running the script.

**Comment by jwerner08** [comment]
Hey all - I put up a pr that adds this functionality: 
https://github.com/firebase/extensions/pull/1687

**Comment by javisank** [comment]
> Hey all - I put up a pr that adds this functionality. If anyone else sees this and needs this functionality in the meantime before it's officially implemented/merged they can use my forked branch `jwerner08:brandfolder_proposal/extensions/firestore_biquery_import_script`. link to pr below point_down
> 
> #1687

Hi, was trying your solution. It doesn't seem to be working the desired way. I think it is still trying to use the same project for firestore and bigquery instances. I think it is related to client instantiation (both firestore and bigquery) with default project (coming from environment variables).

**Comment by jwerner08** [comment]
I ran into this error before and determined it was actually a permissions issue. Here's the steps I used I resolve: 

Steps for using import script from my forked branch: 
1. pull down forked branch: `jwerner08:brandfolder_proposal/extensions/firestore_biquery_import_script`
2. If you haven't already, give bigquery.dataEditor permissions to the service account created via the extension in the GCP project that houses your BQ instance (In my case I gave it dataEditor permissions to a dataset I created).
3. Download the service account key in json file format and put it somewhere locally.
4. Set the `GOOGLE_APPLICATION_CREDENTIALS` environment variable (must be absolute path) `export GOOGLE_APPLICATION_CREDENTIALS="/absolute/path/to/service_account_key.json"
5. Navigate to import directory to run import script.
6. In terminal execute cmd: `run npm build`
7. In terminal execute cmd: `run npm import`
8. Use CLI to specify variables and run. 

☝️ I tested this flow again and was able to import data from a firestore instance into a BQ instance located in a different GCP project.

**Comment by javisank** [comment]
Hi, after a couple of tries, it worked. I couldn't import anything from one project (firestore) to the other (bigquery) until I set the environment variable GCLOUD_PROJECT with firestore instance project Id. I think you should change something because it doesn't take the Firebase Project Id input, rather than the mentioned environment variable. I hope more people test it to check if this is the case or if I am missing something.

**Comment by jwerner08** [comment]
It sounds like local environment variables may supersede any parameter values you specify in the CLI while executing the import script. I'm not the most privy to typescript related stuff, so hopefully the project maintainers gain insight and correct (or document) the implementation so that it's easier for people like us who need the ability to use separate GCP projects for Firestore and BigQuery instances. Glad this helped you though!

**Comment by javisank** [comment]
Yeah, hope they take your work and implement it in the best way. Thank you very much, it unblocked me some tasks I have in my backlog.

**Comment by javisank** [comment]
@jwerner08 wanted to give you a heads up about something.
Reading the code and running the script I discovered that the flag --bigquery-project gets mapped to the variable bigqueryProject. In your code you wrote bigQueryProject. So when running as non-interactive, the variable bigQueryProject is undefined. If you can, change every instance of bigQueryProject to bigqueryProject.

**Comment by profpylons** [comment]
@jwerner08 This is not working for me. It seems to be using the BigQuery project as the source for Firebase: `PERMISSION_DENIED: Cloud Firestore API has not been used in project myproject-datazone-devl before or it is disabled.`

```
$ PROJECT_ID=myproject-datazone-devl  GCLOUD_PROJECT=myproject-datazone-devl FIREBASE_CONFIG=~/.gcloud/myproject-0-devl-firebase-adminsdk-numbers.json npm run import

> @firebaseextensions/fs-bq-import-collection@0.1.18 import
> node ./lib/index.js

? What is your Firebase project ID? myproject-0-devl
? What is your BigQuery project ID? myproject-datazone-devl
? What is the path of the the Cloud Firestore Collection you would like to import from? (This may, or may not, be the same Collection for which you plan to mirror changes.) videos
? Would you like to import documents via a Collection Group query? No
? What is the ID of the BigQuery dataset that you would like to use? (A dataset will be created if it doesn't already exist) devl_staging_app
? What is the identifying prefix of the BigQuery table that you would like to import to? (A table will be created if one doesn't already exist) videos_realtime
? How many documents should the import stream into BigQuery at once? 300
? Where would you like the BigQuery dataset to be located? eu
? Would you like to run the import across multiple threads? No
? Would you like to use the new optimized snapshot query script? Yes
? Would you like to use a local firestore emulator? No
Importing data from Cloud Firestore Collection: videos, to BigQuery Dataset: devl_staging_app, Table: videos_realtime_raw_changelog
Error importing Collection to BigQuery: Error: 7 PERMISSION_DENIED: Cloud Firestore API has not been used in project myproject-datazone-devl before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/firestore.googleapis.com/overview?project=goaldsports-datazone-devl then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.
```

**Comment by jwerner08** [comment]
It looks like you're setting `PROJECT_ID` as the `myproject-datazone-devl` when it should be your Firestore GCP project id `myproject-0-devl`

In `processCollectionGroup()` located at `scripts/import/src/run.ts` you see `ProjectId` and not `bigQueryProjectId` being initialized:
```
async function processCollectionGroup(config: CliConfig): Promise<number> {
  const maxWorkers = Math.ceil(cpus().length / 2);
  const workerPool = pool(__dirname + "/worker.js", {
    maxWorkers,
    forkOpts: {
      env: {
        PROJECT_ID: config.projectId,
        GOOGLE_CLOUD_PROJECT: config.projectId,
        GCLOUD_PROJECT: config.projectId,
        ...process.env,
      },
    },
  });
 ```
 
However it appears you override those values when setting them via env variables: 
`$ PROJECT_ID=myproject-datazone-devl  GCLOUD_PROJECT=myproject-datazone-devl FIREBASE_CONFIG=~/.gcloud/myproject-0-devl-firebase-adminsdk-numbers.json npm run import`

**Comment by profpylons** [comment]
@jwerner08 To be honest, I would hope that doesn't make a difference as the exact projects for firebase and big query come explicitly from the parameters in the config object. eg. Line 69 of `index.js`

```javascript
databaseURL: `https://${projectId}.firebaseio.com`,
```

Also if you look at the initialisation of `FirestoreBigQueryEventHistoryTracker` (line 48) you see that `env.PROJECT_ID` is assumed to be the project of the BigQuery dataset.

```javascript
this.bq.projectId = config.bqProjectId || process.env.PROJECT_ID;
```

There is similar inconsistency in the import script between `index.js` (Project is stated as bqProject) and `run.js` (Project is expected to be fbProject). Although here there's likely some intended/unintended overwriting in `run.js` that gets us where we needed to be. 

Anyway - I'll be down this rabbit hole for a while I think. If I find a way to make this more robust, I'll send you a pull request.

**Comment by jwerner08** [comment]
@profpylons feel free to commit to my forked branch the changes you're suggesting. I'm not a firebase dev, nor a typescript dev, so the pr I submitted may be a little hacky. https://github.com/firebase/extensions/pull/1687

**Comment by Hunnainwali** [comment]
> @javisank
> 
> I ran into this error before and determined it was actually a permissions issue. Here's the steps I used to resolve:
> 
> Steps for using import script from my forked branch:
> 
> 1. Pull down forked branch: `jwerner08:brandfolder_proposal/extensions/firestore_biquery_import_script`
> 2. If you haven't already, give `bigquery.dataEditor` permissions to the service account created via the extension in the GCP project that houses your BQ instance (In my case I gave it dataEditor permissions to a dataset I created).
> 3. Download the service account key in json file format and put it somewhere locally.
> 4. Set the `GOOGLE_APPLICATION_CREDENTIALS` environment variable (must be absolute path) `export GOOGLE_APPLICATION_CREDENTIALS="/absolute/path/to/service_account_key.json"`
> 5. Navigate to import directory to run import script.
> 6. In terminal execute cmd: `npm run build`
> 7. In terminal execute cmd: `npm run import`
> 8. Use CLI to specify parameter values and run.
> 
> ☝️ I tested this flow again and was able to import data from a firestore instance into a BQ instance located in a different GCP project.

I can see firebase-measurement@system.gserviceaccount.com in IAM -> Permissions for Project 

<img width="1025" alt="Screenshot 2023-12-07 at 8 24 54 PM" src="https://github.com/firebase/extensions/assets/64936324/9cc447f8-36ad-4309-a713-54f09b95a3d4">

But not able to locate that in service accounts section.
